nncf/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,561 @@
+from collections import deque
+from copy import deepcopy
+from enum import Enum
+from functools import cmp_to_key
+from itertools import combinations
+from itertools import groupby
+from typing import List, Type
+
+import networkx as nx
+import torch
+
+from nncf.common.graph.graph import NNCFGraph
+from nncf.common.graph.graph_matching import find_subgraphs_matching_pattern
+from nncf.common.graph.operator_metatypes import OperatorMetatype
+from nncf.common.graph.definitions import MODEL_OUTPUT_OP_NAME
+from nncf.torch.dynamic_graph.operation_address import OperationAddress
+from nncf.torch.graph.graph import PTNNCFGraph
+from nncf.torch.hardware.fused_patterns import PT_HW_FUSED_PATTERNS
+from nncf.torch.graph.operator_metatypes import PTDropoutMetatype
+from nncf.torch.graph.operator_metatypes import PTRELUMetatype
+from nncf.torch.graph.operator_metatypes import PTMatMulMetatype
+from nncf.torch.graph.operator_metatypes import PTLinearMetatype
+from nncf.torch.layers import NNCF_MODULES_OP_NAMES
+from nncf.torch.nncf_network import NNCFNetwork
+
+
+IGNORED_NAME_OPERATORS = [*PTDropoutMetatype.get_all_aliases(), MODEL_OUTPUT_OP_NAME]
+
+class BuildingBlock:
+    def __init__(self, start_node, end_node):
+        self.start_node = start_node
+        self.end_node = end_node
+
+class BuildingBlockType(Enum): # only for BERT
+    MSHA = 'MSHA'
+    FF = 'FF'
+    Unknown = 'unknown'
+
+class BuildingBlockInfo:
+    def __init__(self, building_block: BuildingBlock,
+                       op_addresses: List[OperationAddress],
+                       modules: List[torch.nn.Module],
+                       block_type: BuildingBlockType):
+        self.building_block = building_block
+        self.op_addresses = op_addresses
+        self.modules = modules
+        self.block_type = block_type
+
+class SearchGraphNode:
+    """
+    Class describing nodes used in SGraph.
```

Comment:
Refer to entities correctly: `SGraph` -> `SearchGraph`

Reply:
Done.